### PR TITLE
fix: Synapse E2E Test Starvation

### PR DESCRIPTION
**Pull request by ppruthi**
# Summary

This changeset fixes the flakiness in running Synapse E2E Tests due to limited spark pool(s). The changes include:
- Dynamically creating 3 "Large" Apache Spark Pools (Spark v3.2) for each E2E job triggered to run the tests
- Deleting the spark pools created dynamically at the end of each run
- Use of Apache Spark Pool (bigDataPool resource) bicep spec to create each spark pool
- Use RESTful API endpoint of Azure Resource Management for Synapse Spark to create/delete pools
- Modularity to change bicep template for bigDataPool

# Tests

Tested the changes by triggering the E2E tests on a production workspace as well as PPE workspace.

# Dependency changes

None

AB#1810147

### CODE REVIEW

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala [hunk 1]
```diff
@@ -34,12 +34,26 @@ class SynapseTests extends TestBase {
...
+
+  println(s"SynapseTests E2E Test Suite starting. Creating ${expectedPoolCount} Spark Pools...")
+  var (sparkPools, actualPoolCount) = {
+    createSparkPools(expectedPoolCount, subscriptionId, resourceGroupName, workspaceName)
+  }
+  sparkPools = sparkPools.slice(0, actualPoolCount)
+  println("Spark Pools Created: [" + sparkPools.mkString(", ") + "]")
+
+  if (actualPoolCount == 0) {
+    println("Failed to create any Spark Pool for running the tests. Exiting...")
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala:49]: can throw an error here instead

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala [hunk 2]
```diff
@@ -34,12 +34,26 @@ class SynapseTests extends TestBase {
...
-    "e2etstspark32i2",
-    "e2etstspark32i3",
-    "e2etstspark32i4",
-    "e2etstspark32i5")
+  val resourceGroupName = "marhamil-mmlspark"
+  val subscriptionId = "e342c2c0-f844-4b18-9208-52c8c234c30e"
+  val expectedPoolCount = 3
+
+  println(s"SynapseTests E2E Test Suite starting. Creating ${expectedPoolCount} Spark Pools...")
+  var (sparkPools, actualPoolCount) = {
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala:42]: avoid using var

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala [hunk 3]
```diff
@@ -131,8 +137,8 @@ object SynapseUtilities extends HasHttpClient {
   }
 
   def poll(id: Int, livyUrl: String): LivyBatch = {
-    val getStatusRequest = new HttpGet(s"$livyUrl/$id")
-    getStatusRequest.setHeader("Authorization", s"Bearer $Token")
+    val getStatusRequest = new HttpGet(s"${livyUrl}/${id}")
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:140]: nit: you dont always need these parens and intellij might actually complain if they are unecessary, though it depends on the author which they prefer. I dont really care that much either way but food for thought

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala (hunk 2 above)
Comment (mhamilton723) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala:42]: We alkso want to make sure that the pools are independent per build job that is run so its okay not to query for existing pools to re-use but just spin up and tear down dynamically

Comment (ppruthi) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseTests.scala:42]: That is exactly what this PR helps. We create spark pools named "tc<ts-in-millisecond>" (Eg. `tc1653515249401`) for each new e2e run and delete these at the end of graceful shutdowns to ensure per build job we have separate set of pools

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala (hunk 3 above)
Comment (ppruthi) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:140]: I liked the readability with parens -- but its more a style choice --- whatever works for the codebase -- I'd just to be normalized one way or another. Im assuming your preference is not to have 'em ?

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala [hunk 4]
```diff
@@ -230,19 +237,141 @@ object SynapseUtilities extends HasHttpClient {
...
+    }
+    (sparkPool, actualPoolCount)
+  }
 
-    val uri: String = s"https://login.microsoftonline.com/$TenantId/oauth2/token"
+  def deleteSparkPools(sparkPools: Array[String],
+                       subscriptionId: String,
+                       resourceGroupName: String,
+                       workspaceName: String): Boolean = {
+    var deleteAll = true  // Flag to verify if a successful cleanup was executed.
```
[the remark this answers is not in the document]
Comment (mhamilton723) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:317]: This var can be removed by mapping spark pools to a list of statueses and making sure all statuses are Succed

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala [hunk 5]
```diff
@@ -230,31 +222,156 @@ object SynapseUtilities extends HasHttpClient {
...
+    |}
+    |""".stripMargin
+  }
 
-    val uri: String = s"https://login.microsoftonline.com/$TenantId/oauth2/token"
+  def createSparkPools(poolCount: Int = 1,
+                       subscriptionId: String,
+                       resourceGroupName: String,
+                       workspaceName: String): Array[String] = {
+    val sparkPool = new Array[String](poolCount)
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:269]: this is still state that you seem to be mutating down below. consiuder turning your .foreach into a map and returning the spark pool id

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala [hunk 6]
```diff
@@ -230,31 +222,156 @@ object SynapseUtilities extends HasHttpClient {
     }
   }
 
-  def getSynapseToken: String = {
-    val spnKey: String = Secrets.SynapseSpnKey
+  private def getBigDataPoolBicepPayload(bigDataPoolName: String,
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:225]: nit: BigDataPool -> SparkPool

File: core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala [hunk 7]
```diff
@@ -230,31 +222,156 @@ object SynapseUtilities extends HasHttpClient {
...
+          s"""
+             |"Failed to create Apache Spark Pool ${bigDataPoolName} due to Exception: ${e.getMessage}
+             |""".stripMargin
+        )
+      }
+    })
+    sparkPool
+  }
+
+  def deleteSparkPools(sparkPools: Array[String],
```
Comment (mhamilton723) [inline core/src/test/scala/com/microsoft/azure/synapse/ml/nbtest/SynapseUtilities.scala:313]: if this isnt guaranteed to succeed maybe call this tryDeleteSparkPools

**Pull request merged**